python/oneseismic/client/client.py

```diff
@@ -313,44 +233,100 @@ class http_session(requests.Session):
+        if self.tokens is None:
+            return kwargs
+
+        headers = self.tokens.headers()
+        if 'headers' in kwargs:
+            # unpack-and-set rather than just assigning the dictionary, in case
+            # headers() starts returning more than just the Authorization
+            # headers. This puts the power of definition where it belongs, and
+            # keeps http_session oblivious to oneseismic specific header
+            # expectations.
+            #
+            # If users at call-time explicitly set any of these headers,
+            # respect them
+            for k, v in headers.items():
+                kwargs['headers'].setdefault(a)
+        else:
+            kwargs['headers'] = headers
+
+        return kwargs
+
     def get(self, url, *args, **kwargs):
-        """
+        """HTTP GET
+
         requests.Session.get, but raises exception for non-200 HTTP status
-        codes.
+        codes. Authorization headers will be added to the request if
+        http_session.tokens is available.
+
+        This function will respect call-level custom headers, and only use
+        http_session.tokens.headers() if not specified, similar to the requests
+        API [1]_.
 
         Parameters
         ----------
         url : str
             Relative url to the resource, e.g. 'result/<pid>/status'
 
+        Returns
+        -------
+        r : request.Response
+
         See also
         --------
         requests.get
+
+        References
+        ----------
+        .. [1] https://requests.readthedocs.io/en/master/user/advanced/#session-objects
+
+        Examples
+        --------
+        Defaulted and custom authorization:
+        >>> session = http_session(url, tokens = tokens)
+        >>> session.get('/needs-auth')
+        >>> session.get('/needs-auth', headers = { 'Accept': 'text/html' })
+        >>> session.get('/no-auth', headers = { 'Authorization': None })
         """
+        kwargs = self.merge_auth_headers(kwargs)
         r = super().get(f'{self.base_url}/{url}', *args, **kwargs)
         r.raise_for_status()
         return r
 
-class client:
-    def __init__(self, endpoint=None, auth=None, cache_dir=None):
-        cache_dir = cache_dir or XDG_CACHE_HOME
-        config = readconfig(cache_dir)
-        if endpoint is not None:
-            self.endpoint = endpoint
-        else:
-            self.endpoint = config['url']
+    @staticmethod
+    def fromconfig(cache_dir = None):
+        """Create a new session from on-disk config
 
-        if auth is None:
-            auth = azure_auth(cache_dir)
-        self.session = http_session(self.endpoint)
-        self.session.headers.update(auth.token(config))
+        Create a new http_sesssion with parameters and auth read from disk.
+        This is a convenient constructor for most programs and uses outside of
+        testing.
+
+        Parameters
+        ----------
+        cache_dir : path or str, optional
```

Comment:
Perhaps we should specify what this defaults to?

Reply:
I have considered it, but it's really a global default and docs should be derived as such.

Reply:
It's something that's easy to add later anyway, so let's merge for now.